core/state-machine/src/lib.rs

```diff
@@ -245,7 +253,13 @@ pub trait Externalities<H: Hasher> {
 
 	/// Get the change trie root of the current storage overlay at a block with given parent.
 	fn storage_changes_root(&mut self, parent: H::Out) -> Result<Option<H::Out>, ()> where H::Out: Ord;
+}
 
+/// Extra externalities API: pinned to specific active address.
+///
+/// Methods of this trait may have side-effects.  Therefore, it isn't safe
+/// to catch unwinding of any panics that occur during calls to them.
```

Comment:
Can you elaborate what connection do panics and side-effect have here?

Reply:
My thoughts were:
1) we do not update storage directly from `StorageExternalities` methods - instead we're buffering changes && only update storage when block is actually inserted into db. You may think of this as an implementation details (I mean `Ext` + `OverlayChanges`), but imo there's no other way to organize block import - just because if you're writing directly to storage, then any panic will lead to storage corruption. That's why I consider catching panics from within storage-related methods to be `AssertUnwindSafe`;
2) keystore/offchain are singletons, which may have their own global state (I could say for sure that keystore stores some keys in memory). If keystore/offchain method will panic and we will catch this panic, this could leave these objects in some inconsistent state. That's why we can't catch panics from these methods.